src/main/setupInvestHandlers.js

```diff
@@ -166,3 +136,25 @@ export function setupInvestRunHandlers(investExe) {
     });
   });
 }
+
+export function setupInvestLogReaderHandler() {
+  ipcMain.on(ipcMainChannels.INVEST_READ_LOG, async (event, logfile, channel) => {
+    const fileStream = fs.createReadStream(logfile);
+    fileStream.on('error', (err) => {
+      logger.error(err);
+      event.reply(
+        `invest-stdout-${channel}`,
+        `Logfile is missing or unreadable: ${os.EOL}${logfile}`
+      );
+    });
+
+    const rl = readline.createInterface({
```

Comment:
I'm sure this is just my relative node inexperience coming out, but I'm not sure I understand the benefit of having a readline interface atop `fileStream` ... is it just that we're trying to iterate one line at a time rather that whatever chunks `fileStream` would normally read?  Is there a limit to the size of the data that can be sent across the `invest-stdout-*` channel?  Something else?  Sorry for my confusion here!

Reply:
@phargogh That's a great question that I never thought too deeply about. The goal is to read one line at a time because that's how the `LogTab` `markupLine` is designed to work. If you don't mind, could you take a look in `LogTab` and give that design a bit of critical review? I wouldn't want to design this whole `fileStream + readline` stuff around something else that was a flawed idea to begin with.

At some point we'll probably move that `markupLine` feature out of the renderer and into this main process because it's still using one node module (`santizeHtml`) and we ought to eventually remove node integration from the renderer completely. That could mean just substituting in another `ipc` call.

Reply:
Sure!  So, I'm not certain I can speak to whether `markupLine` belongs in the main process or in the renderer at the moment, but your reasoning makes sense to me about that.

The main thing I'm noticing about this design centers around the line-by-line nature of it.  That's not _inherently_ a problem, but the fact that we start from the very beginning with reading from stdout line-by-line is probably less efficient than it could be.  The [Node docs for `createReadStream`](https://nodejs.org/api/fs.html#fs_fs_createreadstream_path_options) suggest that it'll by default read in 16Kb at a time, which is usually a whole lot more than a single line!  The point of deliberately passing around larger chunks is that it tends to reduce iteration overhead, and since each chunk here will be passing through the IPC (where presumably there's some overhead per call), I'd wonder about all of that overhead just generally slowing things down.

So if I were to make an architectural observation, it would be that we might be able to reduce overhead by reading data in larger chunks (the default should be fine) and then only split by newlines where it's needed so that we minimize that communication overhead.  It might not make a difference with just running 1 model, but if we're trying to monitor a bunch of streams all at the same time, I bet that would become noticeable.

And maybe this wouldn't matter at all if we move `markupLine` out of the renderer and we'll need more IPC for that move?  I'm not sure.

And I could of course be totally off base with this ... [my timings](https://gist.github.com/phargogh/00aea868fddcab3def00840da5482dcb) were around simply reading from a file stream, which is fairly simple compared with inter-process communication.  Anyways, I'd be curious what you think about this.

Reply:
We talked about this on the office hours huddle today and decided that there isn't enough of a slowdown to warrant a complete refactor of the line-by-line handling of markup, and we can always try to optimize it later on when we do have a clear slowdown.  @davemfish was also going to try using python's `PYTHONUNBUFFERED=1` environment variable to see how well that works and whether that might be useful and/or faster within the workbench context of reading from stdout.

Reply:
@phargogh here's an update. Adding the `PYTHONUNBUFFERED` env variable seems to work. There's no discernable difference with or without it for the invest runs I've tried, but this still seems like a good use-case for it.

However, reading line-by-line from a file on disk takes too long for even a modest-sized logfile of 1660 lines. It takes a few seconds to load. 

After more thinking, it might be reasonable to only do the markup during active invest runs as the data comes through stdout. Then for loading "Recent Run" logfiles, we just load it all as plaintext with no markup. In the future, if we want to add the markup feature to "Recent Run" logs, we might consider caching the marked-up logfile that we create in-memory during the invest run, and load *that* back later instead of the plaintext logfile. What do you think?

Reply:
Ditching line-by-line would also solve another minor problem: reading logfiles line-by-line is not exactly the same as logger message-by-message. As some messages contain newlines. So there were occasional cases where the pattern-matcher only marked-up the first line of a multi-line logger message. And our regex got more complicated in order to deal with that, especially for Tracebacks.

Reply:
> And our regex got more complicated in order to deal with that, especially for Tracebacks.

Yeah, I was wondering about this case as well.  Remembering back to that PyBay talk we had both attended, would be be worth considering structured logging?

> However, reading line-by-line from a file on disk takes too long for even a modest-sized logfile of 1660 lines. It takes a few seconds to load.
>
> After more thinking, it might be reasonable to only do the markup during active invest runs as the data comes through stdout. 

Oh my, that's a notable delay on a not-terribly long file!  I like your solution of marking up the file, caching it and presenting that to the user.  That feels like a good balance of orthogonality, keeping the UI features in the UI layer and still minimizing the need to recompute the HTML markup on file load.